fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -556,29 +561,57 @@ private void increaseLimit() {
 
         AtomicLong recordsScannedCounter = new AtomicLong();
         // Note: This runs all of the updates in serial in order to not invoke a race condition
-        // in the rank code that was causing incorrect results. If everything were thread safe,
-        // a larger pipeline size would be possible.
-        return cursor.forEachResultAsync(result -> {
+        // in the rank code that was causing incorrect results.
+
+        final AtomicReference<RecordCursorResult<FDBStoredRecord<Message>>> lastResult = new AtomicReference<>(RecordCursorResult.exhausted());
+        final FDBRecordContext context = store.getContext();
+        return AsyncUtil.whileTrue(() -> cursor.onNext().thenCompose(result -> {
+
+            if (!result.hasNext()) {
+                // end of the cursor list
+                if (timer != null) {
+                    timer.increment(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RANGE_BY_COUNT);
+                }
+                lastResult.set(result);
+                return AsyncUtil.READY_FALSE;
+            }
+
             final FDBStoredRecord<Message> rec = result.get();
             empty.set(false);
             if (timer != null) {
                 timer.increment(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_SCANNED);
             }
             recordsScannedCounter.incrementAndGet();
-            if (recordTypes.contains(rec.getRecordType())) {
-                if (timer != null) {
-                    timer.increment(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_INDEXED);
-                }
-                if (syntheticPlan == null) {
-                    return maintainer.update(null, rec);
-                } else {
-                    // Pipeline size is 1, since not all maintainers are thread-safe.
-                    return syntheticPlan.execute(store, rec).forEachAsync(syntheticRecord -> maintainer.update(null, syntheticRecord), 1);
-                }
-            } else {
-                return AsyncUtil.DONE;
+            if (!recordTypes.contains(rec.getRecordType())) {
+                // This record is not our type, swipe left
+                return AsyncUtil.READY_TRUE;
             }
-        }).thenCompose(noNextResult -> {
+            // add this index to the transaction
+            if (timer != null) {
+                timer.increment(FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_INDEXED);
+            }
+            return (syntheticPlan == null ?
+                    // case 1: simple
+                    maintainer.update(null, rec) :
+                    // case 2: synatheticPlan - note that Pipeline size is 1, since not all maintainers are thread-safe.
+                    syntheticPlan.execute(store, rec
+                    ).forEachAsync(syntheticRecord -> maintainer.update(null, syntheticRecord), 1)
```

Comment:
I kind of liked @nblintao's suggestion to remove the ternary operator and make it a real `if` statement for readability, but I suppose it doesn't really matter

Reply:
If the majority claim that A is more readable than B, then more readable it is.